packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts

```diff
@@ -101,26 +106,135 @@ class FluidDataObjectFactory {
+*/
+class PureDataObjectProxy<TObj extends PureDataObject<O, S>, O, S>  {
+    protected instance: TObj | undefined;
+    protected instanceP: Promise<TObj> | undefined;
+
+    public readonly runtime: FluidDataStoreRuntime;
+
+    public static async create<TObj extends PureDataObject<O, S>, O, S>(
+        ctor: new (props: IDataObjectProps<O>) => TObj,
+        context: IFluidDataStoreContext,
+        sharedObjectRegistry: ISharedObjectRegistry,
+        optionalProviders: FluidObjectSymbolProvider<O>,
+        props: S,
+        onDemandInstantiation: boolean)
+    {
+        const instance = new PureDataObjectProxy(
+            ctor,
+            context,
+            sharedObjectRegistry,
+            optionalProviders,
+            props,
+            onDemandInstantiation);
+
+        // if it's a newly created object, we need to wait for it to finish initialization
+        // as that results in creation of DDSs, before it gets attached, providing atomic
+        // guarantee of creation.
+        // WARNING: we can't do the same (yet) for already existing PureDataObject!
+        // This will result in deadlock, as it tries to resolve internal handles, but any
+        // handle resolution goes through root (container runtime), which can't route it back
+        // to this data store, as it's still not initialized and not known to container runtime yet.
+        // In the future, we should address it by using relative paths for handles and be able to resolve
+        // local DDSs while data store is not fully initialized.
+        if (!instance.runtime.existing) {
+            await instance.getInstance();
+        }
+
+        return instance;
+    }
+
+    protected constructor(
+        private readonly ctor: new (props: IDataObjectProps<O>) => TObj,
+        readonly context: IFluidDataStoreContext,
+        private readonly sharedObjectRegistry: ISharedObjectRegistry,
+        private readonly optionalProviders: FluidObjectSymbolProvider<O>,
+        private readonly props: S,
+        onDemandInstantiation: boolean)
+    {
+        // base
+        let runtimeCtor = FluidDataStoreRuntime;
+
+        // request mixin in
+        runtimeCtor = requestFluidDataStoreMixin(
+            runtimeCtor,
+            async (request: IRequest) => (await this.getInstance()).request(request));
+
+        // If factory needs to instantiate PureDataObject, then it means it needs to finish
+        // initialization before anything can be done with runtime. For example, DataObject needs
+        // to provide additional blobs to summary that are not part of DDS.
+        // Please note that ideal solution is for code above to await instanceP, but given that we can't
+        // do it yet for existing data store, we have to do it only when summary is called, i.e. after
+        // data store runtime is fully initialized (from POV of container runtime)
+        if (!onDemandInstantiation) {
```

Comment:
Note: I'm thinking about moving this out from here in next iterations by plumbing through default argument:
```
runtimeCtor: typopeof FluidDataStoreRuntime = FluidDataStoreRuntime
```
That way DataObject can provide a bunch of custom behaviors, without complicating interfaces / implementaitons.

Reply:
Fixed